### DISCUSSION TITLE
Add metadata to route

### DISCUSSION THREAD

**Discussion by MatsiukMykola** [encode/starlette#2259] [Ideas] [answered]
```python
@router.post(
    path=f"/get_books/{CrudEnum.list.value}",
    name=CrudEnum.list, # <- this UGLY, but  don't see any variant to add some metadata to route ?
    description="Get books",
    response_model=BookListResponseSchema,
)
async def get_books(
    session: Session,
    user: User,  # noqa
params: Annotated[
        QBRequest,
        Body(
            openapi_examples={ # <- How easy extract examples based on route?
})]
```

i say to frontend guys - "don't worry see path"
name - for my internal tasks :(

and right now don't understand where to read link (route + function) definition, to extract all examples and run in autotest

are more native way to add any metadata to route ? have easy way to extract

**Comment by Kludex** [comment]
... What?

**Reply by ManiMozaffar** [reply]
Well I actually proto typed it here, please take a look:
https://github.com/tiangolo/fastapi/pull/10158

I think the text might be confusing, but the idea is to have someway to inject data to metadata, and then receive it as dependency, could be quite useful for things like setting up ACL without having to use middleware and "url path" and dirty codes coming after it.

**Reply by Kludex** [reply]
I didn't get what you want. Explain here with a minimal reproducible example, please.

**Reply by ManiMozaffar** [reply]
Ok, let me rephrase it like that, 
Imagine you're writing a ACL configuration for your routers, what would you do? you need to bind a StrEnum, to each Endpoint. so let's say i have
```python
from enum import StrEnum, auto

class EndPoints(StrEnum):
     CREATE_SOMETHING = auto()

@router.post("something/create")
def create_something(acl_func: ... = Depends()):
     await acl_func(endpoint=EndPoints.CREATE_SOMETHING) # shall raise exceptions inside the code if user is not permitted
```
now i need to do that, in all routers right? so it's a duplicate code, everywhere. and it's also a dependency, of each endpoint. I can't call it inside acl_func, because inside acl_func i don't have access to endpoint enum.
now same thing goes for any other metadata that we want to have, there's no control on the router object itself, we can't modify it and attach a metadata, to be used for different purpose, such as ACL Design. but let's say we do as I proto-typed, then

```python
from enum import StrEnum, auto
from fastapi import APIMetaData

class EndPoints(StrEnum):
     CREATE_SOMETHING = auto()


async def no_duplicate_acl(acl_func: ... = Depends(), router_metadata: APIMetaData = Depends():
     await acl_func(endpoint=router_metadata.enum) # shall raise exceptions inside the code if user is not permitted


@router.post("something/create", enum=EndPoints.CREATE_SOMETHING)
def create_something(no_duplicate_acl: ... = Depends()):
     ...

```

**Reply by MatsiukMykola** [reply]
looks like you try implement LDAP
hm... interesting idea, if are any way to inject how dependency from router this will be a really nice idea to extending

I don't try, but maybe this help:
```python
from fastapi import FastAPI, Depends, Request
from fastapi.routing import APIRoute

def router_middleware(request: Request):
    # Access the router from the request scope
    current_route = request.scope.get("router")
    
    # Modify the request context to include the router information
    request.state.router_info = current_route
    
    # Continue the middleware chain
    return None

class MyAPIRoute(APIRoute):
    def get_route_handler(self):
        original_route_handler = super().get_route_handler()
        
        async def custom_route_handler(request: Request):
            # Execute the middleware function before the original route handler
            router_middleware(request)
            
            # Call the original route handler and return its response
            response = await original_route_handler(request)
            return response
        
        return custom_route_handler

app = FastAPI(route_class=MyAPIRoute)

# Define a sample dependency that accesses the router information
def get_router_info(request: Request = Depends()):
    return request.state.router_info

@app.get("/items/", dependencies=[Depends(get_router_info)])
async def read_items():
    return {"message": "Router information injected"}

```

**Reply by Kludex** [reply, marked as answer]
I didn't get the problem, or why you want it in Starlette. You keep using FastAPI, and FastAPI features to explain why you want it here.

If you want something per route, we have middleware per Router, which should be enough.

FastAPI is some releases behind Starlette.

**Comment by MatsiukMykola** [comment]
how possible add data: Any to route object ?

**Reply by alex-oleshkevich** [reply]
Use a decorator on endpoint callable:
```python
@acl(EndPoints.CREATE_SOMETHING)
async def my_endpoint(request): ...
```

Make the decorator to wrap the callable with your ACL logic.

Alternatively:
1. you can use middleware to match the path and apply your logic
2. use `Mount` with middleware to apply your logic to a subset of routes.
